src/core/crypto/aes_ccm.cpp

```diff
@@ -42,14 +42,31 @@
 namespace ot {
 namespace Crypto {
 
+void AesCcm::SetKey(otCryptoKey *aKey)
+{
+    mEcb.SetKey(aKey);
+}
+
 void AesCcm::SetKey(const uint8_t *aKey, uint16_t aKeyLength)
 {
-    mEcb.SetKey(aKey, CHAR_BIT * aKeyLength);
+    otCryptoKey aCryptoKey;
+
+    aCryptoKey.mKey       = aKey;
+    aCryptoKey.mKeyLength = aKeyLength;
+    aCryptoKey.mKeyRef    = 0;
+
+    mEcb.SetKey(&aCryptoKey);
 }
 
 void AesCcm::SetKey(const Mac::Key &aMacKey)
 {
-    SetKey(aMacKey.GetKey(), Mac::Key::kSize);
+    otCryptoKey aCryptoKey;
+
+    aCryptoKey.mKey       = aMacKey.GetKey();
+    aCryptoKey.mKeyLength = Mac::Key::kSize;
+    aCryptoKey.mKeyRef    = aMacKey.GetKeyRef();
```

Comment:
@abtink Yes this was intended and works fine with PSA, as I usually test using PSA before updates. This was intended to avoid fragmentation of code by using ifdefs. You are right `otMacKeyMaterial` is a union and in case of PSA, `GetKey` never returns `nullptr`. That is why, we are not validating `mKey` or `mKeyRef` in AES modules. we pass the otCryptoKey variable after initialising `mKeyRef` and `mKey`. And based on `OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE`, either one of them will be invalid. It is upto the Crypto platform abstraction to use the correct member based on the implementation supported.

For example, the default mbedTLS implementation has this:
```
OT_TOOL_WEAK otError otPlatCryptoAesSetKey(void *aContext, size_t aContextSize, otCryptoKey *aKey)
{
    otError              error    = OT_ERROR_NONE;
    mbedtls_aes_context *mContext = static_cast<mbedtls_aes_context *>(aContext);

    VerifyOrExit(aContextSize >= sizeof(mbedtls_aes_context), error = OT_ERROR_FAILED);
    VerifyOrExit((mbedtls_aes_setkey_enc(mContext, aKey->mKey, (aKey->mKeyLength * CHAR_BIT)) == 0),
                 error = OT_ERROR_FAILED);

exit:
    return error;
}
```

As we know we want only the `mKey`, we ignore the mKeyRef. And in my PSA implementation I ignore the mKey, as I know we support PSA.

Reply:
> And based on `OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE`, either one of them will be invalid. It is up to the Crypto platform abstraction to use the correct member based on the implementation supported.

I am not sure if this is a good approach/model (even if it may work for now with the current platform implementation). 
I would recommend not using it.
- In the definition of `otCryptoKey` we have that `mKeyRef` is valid only when `mKey` is `NULL`, which would not be case here: 
```c++
struct otCryptoKey
{
    const uint8_t *mKey;       ///< Pointer to the buffer containing key. NULL indicates to use `mKeyRef`.
    uint16_t       mKeyLength; ///< The key length in bytes (applicable when `mKey` is not NULL).
    uint32_t       mKeyRef;    ///< The PSA key ref (requires `mKey` to be NULL).
};
```
- Basically I think here we are not  following the documentation of how `otCryptoKey` should be used by platform layer (and can set `mKey` as a non-`NULL` even when we intend to use `mKeyRef` and then expect platform layer to be aware of `OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE` config and use it instead to deterime whethere to use `mKey` or `mKeyRef`).
- I think it is better to not assume that platform layer code would see the same set of OT configs. The `OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE` is really only intended for the OT core.
- I notice that in the code from the PR in other places, we follow the pattern of setting `mKey` to `nullptr` when we want to use `mKeyRef`. For example we have this patten in a few places in `key_manager` (e.g., in `ComputeKeys()` and `ComputeTrelKey()`)
```c++
void KeyManager::ComputeKeys(uint32_t aKeySequence, HashKeys &aHashKeys)
{
    Crypto::HmacSha256 hmac;
    uint8_t            keySequenceBytes[sizeof(uint32_t)];
    otCryptoKey        aKeyMaterial;

#if OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE
    aKeyMaterial.mKey    = nullptr;
    aKeyMaterial.mKeyRef = mNetworkKeyRef;
#else
    aKeyMaterial.mKey       = mNetworkKey.m8;
    aKeyMaterial.mKeyLength = sizeof(mNetworkKey.m8);
    aKeyMaterial.mKeyRef    = 0;
#endif
```
- I think it may be better to be consistent and use the same pattern everywhere.

Reply:
Agreed I will change this.

Reply:
An idea here since we may end up using similar patterns in different places is to think if we can encapsulate them into a helper methods and reuse them to simplify the code.

For example, we can add a method on `Mac::KeyMaterial` class, that converts the `Mac::KeyMaterial` instance to a `otCryptoKey` (populates the `otCryptoKey` from the `Mac::key`):
```c++
void Mac::KeyMaterial::ConvertTo(otCryptoKey &aCryptoKey) const
{
#if OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE
    aCryptoKey.mKey    = nullptr;
    aCryptoKey.mKeyRef = GetKeyRef();
#else
    aCryptoKey.mKey       = GetKey();
    aCryptoKey.mKeyLength = kSize;
    aCryptoKey.mKeyRef    = 0;
#endif
}
```
--------

We may have other code patterns that we can encapsulate as helper methods and re-use.

One pattern that I see we use in bunch of places in `KeyManager` is the following:
```c++
void KeyManager::UpdateKeyMaterial(void)
{
    // .... calculate some `HashKey`
#if OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE
    otMacKeyRef keyRef = 0;

    error = otPlatCryptoImportKey(&keyRef, OT_CRYPTO_KEY_TYPE_AES, OT_CRYPTO_KEY_ALG_AES_ECB,
                                  (OT_CRYPTO_KEY_USAGE_ENCRYPT | OT_CRYPTO_KEY_USAGE_DECRYPT),
                                  OT_CRYPTO_KEY_STORAGE_VOLATILE, cur.mKeys.mMacKey.GetKey(), cur.mKeys.mMacKey.kSize);

    OT_ASSERT(error == kErrorNone);
    cur.mKeys.mMacKey.Clear();
    cur.mKeys.mMacKey.mKeyMaterial.mKeyRef = keyRef;

    keyRef = 0;
    CheckAndDestroyStoredKey(mMleKey.mKeyMaterial.mKeyRef);

    error = otPlatCryptoImportKey(&keyRef, OT_CRYPTO_KEY_TYPE_AES, OT_CRYPTO_KEY_ALG_AES_ECB,
                                  (OT_CRYPTO_KEY_USAGE_ENCRYPT | OT_CRYPTO_KEY_USAGE_DECRYPT),
                                  OT_CRYPTO_KEY_STORAGE_VOLATILE, cur.mKeys.mMleKey.GetKey(), cur.mKeys.mMleKey.kSize);

    OT_ASSERT(error == kErrorNone);
    cur.mKeys.mMleKey.Clear();
    mMleKey.mKeyMaterial.mKeyRef = keyRef;
#else
    mMleKey                 = cur.mKeys.mMleKey;
#endif
```


Basically we calculate/derive a new literal  key by hashing (`HashKey`), then:
- We want to convert and save it as a `Mac::KeyMatrial` (which is the `union` type `otMacKeyMaterial)
- We check if `OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE` is not enabled
   -  We copy the literal key as `Mac::KeyMaterial` and use it directly.
- Otherwise, if `OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE` is enabled, then
   - We need to imporat the literal key to ITS (`otPlatCryptoImportKey()` and get a key ref associated with it.
   - Then set the `mKeyRef` in the `Mac::KeyMaterial`

We can consider adding a helper method for this in `Mac::KeyMaterial` like this:
```c++
// This represents a literal key.
class Key : public otMacKey
{ ... };

// This represents a key material (union or either litral key or a key ref)
class KeyMaterial : public otMacKeyMaterial
{ ... };

// Sets the `KeyMaterial` from a given literal key.
void KeyMaterial::SetFrom(const Key &aKey)
{
    Clear();

#if  OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE
    {
        // Import key into ITS and remember the key ref
        Error error = otPlatCryptoImportKey(&mKeyRef, OT_CRYPTO_KEY_TYPE_AES, OT_CRYPTO_KEY_ALG_AES_ECB,
                                           (OT_CRYPTO_KEY_USAGE_ENCRYPT | OT_CRYPTO_KEY_USAGE_DECRYPT),
                                           OT_CRYPTO_KEY_STORAGE_VOLATILE, aKey.m8, kSize);

        OT_ASSERT(error == kErrorNone);
    }
#else
    // Use the literal key
    mKey = aKey;
#endif
}
```


I think such encapsulations can help simply  the code and also limit the number of places where we need to do `#if OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE` checks.